### PR TITLE
Add Liveness and Readiness Probes

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

Add support for `httpGet` and `tcpSocket` readiness and liveness probes. In order to achive that, the follwoing things were done in this PR:

* Add an extra `kubevirt-infra` pod which is now responsible to indicate a successful libvirt-start, to free the readiness probe on the compute container.
* Add `httpGet` and `tcpSocket` probes to the API for `spec.readinessProbe` and `spec.livenessProbe`. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add httpGet and tcpSocet livenessProbe and readinessProbe support.
```

### CODE REVIEW

**Comment by rmohr** [comment]
/hold cancel

Ready for review

**Comment by mfranczy** [comment]
Tests are failing:
```
[Fail] Probes for liveness should not fail the VMI [It] with working TCP probe and tcp server 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:2565

[Fail] Probes for liveness should not fail the VMI [It] with working HTTP probe and http server 
/root/go/src/kubevirt.io/kubevirt/tests/utils.go:2582
```

**Comment by rmohr** [comment]
> PR looks nice. Btw. sometimes mentioned tests are failing locally as well.

@mfranczy thx. Maybe I over-optimized the tests a little bit to make them faster. Changed the liveness probe initial delay. Let's see.

**Comment by rmohr** [comment]
tests regarding to this PR look good, but rerunning any way.

**Comment by rmohr** [comment]
ci test please

**Comment by rmohr** [comment]
/hold

There seem to be issues in the functional tests (most likely container indices changed with the new infra container).

**Comment by rmohr** [comment]
/hold cancel

Fixed. Let's see.

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -534,27 +580,26 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
...
+		VolumeMounts:   volumeMounts,
+		ReadinessProbe: defaultReadinessProbe,
+		Resources:      resources,
+		Ports:          ports,
+	}
+
+	if vmi.Spec.ReadinessProbe != nil {
+		container.ReadinessProbe = copyProbe(vmi.Spec.ReadinessProbe)
+		container.ReadinessProbe.InitialDelaySeconds = container.ReadinessProbe.InitialDelaySeconds + LibvirtStartupDelay
+	} else {
```
Comment (cynepco3hahue) [inline pkg/virt-controller/services/template.go]: you already set it, when you create container struct, so you can drop `else`

**Comment by cynepco3hahue** [comment]
@rmohr I can see some suspicious errors:
```
Unexpected Warning event received: testvmi2k7kqkw92x5nhjvpnp5rrqm4nqnxxf55r2p6cf4l6hc5r5cl,25020d9d-ef46-11e8-bc3d-525500d15501: Error on handing over pod: Operation cannot be fulfilled on pods "virt-launcher-testvmi2k7kqkw92x5nhjvpnp5rrqm4nqnxxf55r2p6cfgfqj": the object has been modified; please apply your changes to the latest version and try again
```

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by rmohr** [comment]
/hold

Still something wrong with tests and #1672 should go first.

**Comment by rmohr** [comment]
for reference: http://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/2619/artifact/ci_build_summary.html

**Comment by rmohr** [comment]
ci test please

**Comment by rmohr** [comment]
> just thinking loud: would it make sense to only use the new readyContainer when we actually have a custom readiness probe on the VMI defined? Otherwise we have 2 containers with the same probe, which might be a bit waste of resources.

@slintes yes good point. Let me look into it.

**Comment by rmohr** [comment]
ci test please

File: pkg/virt-controller/services/template.go (hunk 1 above)
Comment (slintes) [reply pkg/virt-controller/services/template.go]: ping @rmohr

**Comment by rmohr** [comment]
for reference: https://jenkins.ovirt.org/blue/organizations/jenkins/kubevirt_kubevirt_standard-check-pr/detail/kubevirt_kubevirt_standard-check-pr/2685

Indications that memory might be tight on the nodes. Increasing the memory.

**Comment by michalskrivanek** [comment]
@fabiand @rmohr since @mmirecki is just adding qemu-ga polling, did you consider using it as a "default" implementation of liveness and/or readiness probes for the guest? (readiness would need the polling interval to be a bit smarter to handle boots better (i.e. more aggressive for first few minutes)).

**Comment by fabiand** [comment]
Well, we have the ability to detect a guest. Thus I could imagine that if we detect it, that we automatically use it to do the liveness probes.

**Comment by rmohr** [comment]
> is just adding qemu-ga polling, did you consider using it as a "default" implementation of liveness and/or readiness probes for the guest?

@michalskrivanek I would say "maybe". I would rather have envisioned an explicit `guestAgent` readiness probe (like `httpGet` and `tcpSocket`) but I would not want to rule it out right now to use it as default if we detect the guest agent.

**Comment by fabiand** [comment]
+1

**Comment by rmohr** [comment]
for reference: https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/2696/artifact/ci_build_summary.html

**Comment by rmohr** [comment]
for reference: http://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/2705/artifact/ci_build_summary.html

**Comment by rmohr** [comment]
For reference: https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/2716//artifact/check-patch.openshift-3.11-crio-release.el7.x86_64/mock_logs/script/stdout_stderr.log.

The k8s lanes pass completely, execept the migration test, which fails because migrations are broken. Increasing the test timeout for the openshift test-lane, we were very close to the 3 hour boundary before. With the new tests, we are slightly above it. The openshift tests still show interesting errors, but which are not related to this PR.

**Comment by rmohr** [comment]
For reference: https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/2720/artifact/ci_build_summary.html

Three different tests failing in some lanes:

```
[Fail] Configurations VirtualMachineInstance definition with guestAgent [It] VMI condition should signal agent presence 
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:488
[Fail] Probes for readiness should succeed [It] with working HTTP probe and http server 
/root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:69
[Fail] Networking VirtualMachineInstance with custom PCI address [It] should configure custom Pci address 
/root/go/src/kubevirt.io/kubevirt/tests/vmi_networking_test.go:536
```

Increasing timeouts.

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
@rmohr I think we pretty close
Can you please check test failure
```
â€¢ Failure [117.695 seconds]
Probes
/root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:18
  for readiness
  /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:28
    should succeed
    /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
      with working HTTP probe and http server [It]
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46

      Timed out after 60.000s.
      Expected
          <v1.ConditionStatus>: False
      to equal
          <v1.ConditionStatus>: True

      /root/go/src/kubevirt.io/kubevirt/tests/probes_test.go:69
------------------------------
STEP: Specifying a VMI with a readiness probe
STEP: Starting the server inside the VMI
{"component":"tests","level":"info","msg":"[{1  \r\n$  [$ ]} {3 screen -d -m nc -klp 1500 -e echo -e \"HTTP/1.1 200 OK\\n\\nHello World!\"\r\n$  [$ ]} {5 echo $?\r\n0\r\n$  [0]}]","pos":"utils.go:2660","timestamp":"2018-12-02T16:39:34.331588Z"}
STEP: Checking that the VMI will be marked as ready to receive traffic
â€¢ [SLOW TEST:77.848 seconds]
```

I think it can be related to some general network instability.

**Comment by cynepco3hahue** [comment]
Last failures does not relate to the PR, and we do not have probes tests failures anymore, so I intend to merge this PR :smile:

**Pull request merged**